Node Color Frontend

ravi-kumar-pilla opened the pull request:
## Description

Partially resolves #2480 

## Development notes

- Implemented frontend support for `node_extras.styles` so node-level appearance can be driven by data from the backend.
- Added `processNodeStyles` helper in `src/components/draw/utils/draw-utils.js`
- Added a useEffect that iterates node DOM groups and applies processed styles when `node.extras.styles` exists:
  - Converts JS-style camelCase keys into CSS property names (strokeWidth → stroke-width) before applying.
  - Applies background-related properties (keys that include background or keys in ['fill','stroke','strokeWidth','opacity']) to the `.pipeline-node__bg` element using selection.style(cssProperty, value, 'important').
  - Applies text/font/color-style properties (keys including text, font, or the 'color' key) to the `.pipeline-node__text` and `.pipeline-node__icon` elements. For color specifically, sets fill instead of the property name.
  - Marks nodes with applied custom styles by adding the `kedro-viz-custom-styled` class to the node group.
  - The effect runs on changes to theme and nodes.
- Added a lightweight selector in `src/selectors/nodes.js` to surface node extras to the frontend
- `getNodeData` now accepts `nodeExtras` and uses it when producing node data - enabling the renderer to read `node.extras.styles`.
- Updated store normalization `src/store/normalize-data.js` to include extras in the initial pipeline state

## QA notes

- To QA along with backend, follow the instructions for backend setup in #2487 
- For frontend, checkout this branch
- Start both backend server (`make run`) and frontend (`npm start`)
- Load a graph where nodes include `node_extras.styles` in the `main` API. For this make sure your `demo_project` has `.viz/styles.json`

**NOTE:** If you do not want backend changes and still want to see color effects in the frontend, you can initialize the store with a default like below -

At line 165 in `src/store/normalize-data.js` provide a default node_extras which applies to all nodes.

```js
state.node.extras[id] = node.node_extras || {
    stats: {
      rows: 77096,
      columns: 5,
      file_size: 1810602
    },
    styles: {
      color: "white",
      themes: {
        dark: {
          fill: "red"
        },
        light: {
          fill: "green"
        }
      }
    }
  }
```
- Alternatively, I created a combined PR for testing here - #2489 

**Verify:**
- Nodes with `node_extras.styles` show the expected background, stroke, opacity, fill and text/icon color/font changes.
- Theme-specific overrides (if node.extras.styles.themes) apply when switching theme (dark/light).
- Nodes with styles get the `kedro-viz-custom-styled` class on their node group in the DOM.
- color style applies to text/icon fill as intended.
- No JS console errors are thrown when nodes without styles are rendered.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

ravi-kumar-pilla commented on src/components/draw/draw-nodes.jsx:
```diff
@@ -296,7 +298,64 @@ export function DrawNodes({
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [nodes, orientation]);
 
+  // --- Apply theme-based node styling ---
```
Hi @Huongg , I felt DrawNodes is where the flowchart nodes are being rendered. The styles are applied for each node, so the loop is needed. Please suggest an alternative as I haven't worked on React for a while. I see this is how the animations and other things work on flowchart in this component. 

I separated the useEffect so this logic is not triggered everytime but just when there is a theme change or nodes change. I could not think of any other efficient way to do this.

> And also how about in the Workflow component? Does the custom styling gets applied there too? or just in Flowchart only?

Good question. I just saw we are using DrawNodes in workflow component too. So this might affect workflow component which should not be the case. Let me see what I can do. In the meantime, please suggest a better way if any. Thank you

Huongg replied:
Hey Im thinking of something like this

flowchart.jsx

```

 /**
   * Create node style overrides map for theme-based styling
   */
  createNodeStyleOverrides = (nodes, theme) => {
    const nodeStyleOverrides = {};
    
    nodes.forEach(node => {
      if (node.extras && node.extras.styles && Object.keys(node.extras.styles).length > 0) {
        const processedStyles = processNodeStyles(node.extras.styles, theme);
        if (processedStyles) {
          nodeStyleOverrides[node.id] = processedStyles;
        }
      }
    });
    
    return nodeStyleOverrides;
  };


 return {
	……

    // Create node style overrides for theme-based styling
    const nodeStyleOverrides = this.createNodeStyleOverrides(nodes, theme);

   <DrawNodes
	…..
        nodeStyleOverrides={nodeStyleOverrides}
   />

}


```


Then in draw-nodes.jsx

```
export function DrawNodes({
  ….
 nodeStyleOverrides = {},
   
}) => {
   // --- Apply theme-based node styling ---
  useEffect(() => {
    const selections = getNodeSelections(groupRef, nodes);
    if (!selections) {
      return;
    }
    const { allNodes } = selections;
    
    // Apply custom styling to nodes with style overrides
    allNodes.each(function (node) {
      const nodeGroup = select(this);
      const nodeStyles = nodeStyleOverrides[node.id];
      
      if (nodeStyles) {
        Object.entries(nodeStyles).forEach(([key, value]) => {
          const cssProperty = key.replace(/([A-Z])/g, '-$1').toLowerCase();

          // Background properties
          if (
            key.toLowerCase().includes('background') ||
            ['fill', 'stroke', 'strokeWidth', 'opacity'].includes(key)
          ) {
            const bgElement = nodeGroup.select('.pipeline-node__bg');
            bgElement.style(cssProperty, value, 'important');
          }

          // Text properties
          else if (
            key.toLowerCase().includes('text') ||
            key.toLowerCase().includes('font') ||
            ['color'].includes(key)
          ) {
            const textElement = nodeGroup.select('.pipeline-node__text');
            const iconElement = nodeGroup.select('.pipeline-node__icon');

            [textElement, iconElement].forEach((element) => {
              if (key === 'color') {
                element.style('fill', value, 'important');
              } else {
                element.style(cssProperty, value, 'important');
              }
            });
          }
        });
        nodeGroup.classed('kedro-viz-custom-styled', true);
      }
    });
  }, [nodeStyleOverrides, nodes]);

}

Huongg replied:
and i expect nodeStyleOverrides to return this format

```
{
  "node-id-1": {
    "color": "#ffffff",
    // ... other processed CSS properties
  },
  "node-id-2": {
    "fill": "#00ff00",
    // ... other processed CSS properties
  },
  // ... more node IDs
}
```

jitu5 commented:
@ravi-kumar-pilla I observed one small thing when testing it locally,
With below style setting,
```
  state.node.extras[id] = node.node_extras || {
    stats: {
      rows: 77096,
      columns: 5,
      file_size: 1810602
    },
    styles: {
      themes: {
        dark: {
          fill: "red",
          color: "blue",
        }
      }
    }
  }
```

In dark theme it correctly applies red/blue colours, but if you check theme from button to light, colour remain same instead of default. But it works perfectly on reload. I am not sure if its a big problem to solve.  

![lsp-validqqqq1](https://github.com/user-attachments/assets/4e343bae-c454-4575-a181-cc9ae069d448)

ravi-kumar-pilla commented:
Hi @rashidakanchwala and @jitu5, 

Thank you for pointing this out. There was a bug in the FE logic where I was not resetting the styles applied before reapplying when the theme is switched. I fixed it here - https://github.com/kedro-org/kedro-viz/pull/2488/commits/74b8eb07572d5c743137f1ef46b21853b0d3dd01

When I was debugging the above issue, I found the `createNodeStyleOverrides` method [here](https://github.com/kedro-org/kedro-viz/pull/2488/commits/74b8eb07572d5c743137f1ef46b21853b0d3dd01#diff-7c7aba87a2372c7ae12bcfdbdacad000cbe1b688dceb472107df95f00dbf941fR775) was called multiple times during rendering. I fixed it using a conditional check to call the method when the styles change - https://github.com/kedro-org/kedro-viz/pull/2488/commits/74b8eb07572d5c743137f1ef46b21853b0d3dd01#diff-7c7aba87a2372c7ae12bcfdbdacad000cbe1b688dceb472107df95f00dbf941fR797

Please have a look and let me know if there are any issues. Thank you

rashidakanchwala commented:
It works for me now! are we going to merge these two do one last QA before we merge to main?

jitu5 reviewed (approved):
Thanks

ravi-kumar-pilla commented:
Hi Team, Now that the backend is merged, this branch contains both the changes. I tested it locally and it works. If you want to give a try please do so. Another question I have is, since there is no way to disable colors for now, 

1. I renamed `styles.json` to `styles.example.json` so our demo project will not have styles by default. But if we want to demo custom colors, we can rename it to `styles.json`
2. The other option would be to have styles only in light theme which is rarely used, and we can have no colors in dark theme so our regular experience in dark theme stays the same.

Please let me know your thoughts. @jitu5 @rashidakanchwala . Thank you
